groovy-language-server/src/main/java/com/palantir/groovylanguageserver/GroovycWrapper.java

```diff
@@ -116,6 +131,61 @@ public Path getWorkspaceRoot() {
     }
 
     @Override
+    public Map<String, Set<SymbolInformation>> getReferences() {
+        return references;
+    }
+
+    @Override
+    public Set<Location> findReferences(TextDocumentPositionParams params) {
+        Set<SymbolInformation> symbols = fileSymbols.get(params.getTextDocument().getUri());
+        if (symbols == null) {
+            return Sets.newHashSet();
+        }
+        Set<SymbolInformation> foundSymbolInformations = symbols.stream()
+                .filter(s -> Ranges.isValid(s.getLocation().getRange())
+                        && Ranges.contains(s.getLocation().getRange(), params.getPosition())
+                        && (s.getKind() == SymbolKind.Class || s.getKind() == SymbolKind.Interface))
+                .collect(Collectors.toSet());
+
+        if (foundSymbolInformations.isEmpty()) {
+            return Sets.newHashSet();
+        }
+
+        Optional<SymbolInformation> foundSymbol;
+        if (foundSymbolInformations.size() == 1) {
+            foundSymbol = Optional.of(Iterables.getOnlyElement(foundSymbolInformations));
+        } else {
+            foundSymbol = reduceFoundSymbolsToMostSpecific(foundSymbolInformations, params.getPosition());
+        }
+
+        if (!foundSymbol.isPresent()) {
+            return Sets.newHashSet();
+        }
+
+        String symbolName = foundSymbol.get().getName();
+        if (references.get(symbolName) != null) {
+            return references.get(symbolName).stream().map(s -> s.getLocation())
+                    .filter(l -> Ranges.isValid(l.getRange())).collect(Collectors.toSet());
+        }
+
+        return Sets.newHashSet();
+    }
+
+    private Optional<SymbolInformation> reduceFoundSymbolsToMostSpecific(Set<SymbolInformation> symbols,
```

Comment:
`GroovycWrapperTest#testFindReferences_edgeCases` has an example for this.
Basically if you have this code:
```
class A {
B myB
class B {}
}
```
Let's say you call `findReferences` with position line 3, char 7. (exactly on top of the defined class B)
When you are iterating through your symbols, you will find two whose ranges contain this position. These symbols will be the class symbols for A and B, because their ranges contain their whole block.
So here in this function we look for the symbol who contains the position only in the first line of their range. Therefore we only find B, since its range starts on line 3.